### PR TITLE
#5592: Interleaved2ShardedPartialOp, Sharded2InterleavedPartialOp, Matmul1d height sharding + padding fixes

### PR DESCRIPTION
**Pull request by pavlepopovic**
Add InterleavedToShardedPartialOp, and ShardedToInterleavedPartialOps, and tests for them.
Currently share a lot of code with InterleavedToSharded ops, will clean up tomorrow before merging.

The ops provide functionality as InterleavedToSharded/ShardedToInterleaved ops, with a difference that you provide into how many slices the tensor will be sliced into before being converted to sharded, and what slice is being converted to sharded.

A sequence of ops running on slices, with all intermediate results being sharded, is given in test_falcon7b_attnention_sliced, and it represents attention sequence from falcon7b prefill model. (Without bcast op, as it doesn't support shariding ATM)

To get that sequence running, it was needed to remove asserts from matmul 1d height sharded that forbids pading (added test for padded matmul1d)

### CODE REVIEW

**Review by eyonland — APPROVED** [review]

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp [hunk 1]
```diff
@@ -927,9 +927,7 @@ void Matmul::validate(
                         uint32_t per_core_M = program_config.per_core_M;
                         auto shard_shape = input_tensor_a.shard_spec().value().shape;
 
-                        // No padding
-                        TT_FATAL(M % per_core_M == 0);
-                        TT_FATAL(M / per_core_M == input_tensor_a.shard_spec().value().grid.num_cores());
+                        TT_FATAL(div_up(M, per_core_M) == input_tensor_a.shard_spec().value().grid.num_cores());
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp:930]: Did you try out a few cases for padded? Does it work?

**Review by TT-BrianLiu — APPROVED** [review]

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp (hunk 1 above)
Comment (pavlepopovic) [reply tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp:930]: Yes, I've added 1 unit test in test_sharded.py, and another test which represents falcon7b attention matmuls where we are hitting this issue in the first place if we try sharding. (this is in test_falcon_matmuls file)

File: tt_eager/tt_dnn/op_library/sharded/sharded_op.cpp [hunk 2]
```diff
@@ -67,9 +67,9 @@ operation::ProgramWithCallbacks Sharded::create_program(
     auto& output_tensor = output_tensors.at(0);
 
     if (this->sharded_op_type == ShardedOpType::InterleavedToSharded) {
-        return interleaved_to_sharded_multi_core(input_tensor, output_tensor);
+        return interleaved_to_sharded_multi_core(input_tensor, output_tensor, 1, 0);
```
Comment (miacim) [inline tt_eager/tt_dnn/op_library/sharded/sharded_op.cpp:70]: Do you really need 1, 0 to be specified here? It looks like some magic numbers, bit confusing..

File: tt_eager/tt_dnn/op_library/sharded/multi_core/sharded_op_multi_core.cpp [hunk 3]
```diff
@@ -379,6 +390,18 @@ operation::ProgramWithCallbacks sharded_to_interleaved_multi_core(const Tensor&
     tt_metal::SetRuntimeArgs(program, unary_reader_kernel_id, all_cores, {num_units_per_shard});
 
     uint32_t curr_idx_h = 0, curr_idx_w = 0;
+    if (num_slices > 1) {
```
Comment (miacim) [inline tt_eager/tt_dnn/op_library/sharded/multi_core/sharded_op_multi_core.cpp:393]: Duplicate code. Only difference is tensor (input/output). To me it looks better to have function that will return start_idx_h.

File: tt_eager/tt_dnn/op_library/sharded/multi_core/sharded_op_multi_core.cpp [hunk 4]
```diff
@@ -467,15 +490,22 @@ operation::ProgramWithCallbacks sharded_to_interleaved_multi_core(const Tensor&
...
                                                    const std::vector<Tensor>& input_tensors,
                                                    const std::vector<std::optional<const Tensor>>&,
                                                    const std::vector<Tensor>& output_tensors) {
         auto src_buffer = input_tensors.at(0).buffer();
 
-        auto dst_buffer = output_tensors.at(0).buffer();
+        Buffer* dst_buffer = nullptr;
+        if (num_slices > 1) {
+            // If we have num_slices > 1, it means that our op is S->I partial.
+            // And currently we store output tensors there as input[1]
```
Comment (miacim) [inline tt_eager/tt_dnn/op_library/sharded/multi_core/sharded_op_multi_core.cpp:504]: Why we are storing tensors in inputs_tensors.at(1)?

File: tt_eager/tt_dnn/op_library/sharded/sharded_op.cpp (hunk 2 above)
Comment (pavlepopovic) [reply tt_eager/tt_dnn/op_library/sharded/sharded_op.cpp:70]: Can put them in default args

File: tt_eager/tt_dnn/op_library/sharded/multi_core/sharded_op_multi_core.cpp (hunk 4 above)
Comment (pavlepopovic) [reply tt_eager/tt_dnn/op_library/sharded/multi_core/sharded_op_multi_core.cpp:504]: That's how it's done in a similar op when you need to take in optional output tensor

File: tt_eager/tt_dnn/op_library/sharded/multi_core/sharded_op_multi_core.cpp (hunk 3 above)
Comment (pavlepopovic) [reply tt_eager/tt_dnn/op_library/sharded/multi_core/sharded_op_multi_core.cpp:393]: Will do

File: tt_eager/tt_dnn/op_library/sharded_partial/sharded_op_partial.hpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            if constexpr (std::is_same_v<GridType, CoreCoord>) {
+                grid_size = grid;
+                uint32_t num_cores = 0;
+                uint32_t total_height = input_tensor.volume() / input_tensor.get_legacy_shape()[-1];
+                total_height /= num_slices;
+
+                uint32_t total_width = input_tensor.get_legacy_shape()[-1];
+                switch (shard_scheme) {
+                    case TensorMemoryLayout::HEIGHT_SHARDED: num_cores = div_up(total_height, shard_shape[0]); break;
+                    case TensorMemoryLayout::WIDTH_SHARDED: num_cores = div_up(total_width, shard_shape[1]); break;
```
[the remark this answers is not in the document]
Comment (pavlepopovic) [reply tt_eager/tt_dnn/op_library/sharded_partial/sharded_op_partial.hpp:82]: Will add

**Review by pavlejosipovic — APPROVED** [review]

File: tt_eager/tt_dnn/op_library/sharded_partial/sharded_op_partial.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (pavlepopovic) [reply tt_eager/tt_dnn/op_library/sharded_partial/sharded_op_partial.hpp:82]: added

**Comment by pavlepopovic** [comment]
Pipeline:
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8346062786

**Pull request merged**